src/main/java/org/apache/sysds/runtime/util/DoubleBufferingOutputStream.java

```diff
@@ -43,42 +42,49 @@ public DoubleBufferingOutputStream(OutputStream out, int num, int size) {
 		super(out);
 		if(size <= 0)
 			throw new IllegalArgumentException("Buffer size <= 0.");
-		if( size%8 != 0 )
+		if(size % 8 != 0)
 			throw new IllegalArgumentException("Buffer size not a multiple of 8.");
 		_buff = new byte[num][size];
 		_locks = new Future<?>[num];
-		for(int i=0; i<num; i++)
+		for(int i = 0; i < num; i++) // fill the futures to avoid null pointers.
 			_locks[i] = ConcurrentUtils.constantFuture(null);
 	}
 
 	@Override
 	public void write(int b) throws IOException {
-		throw new IOException("Not supported"); 
+		throw new IOException("Not supported");
 	}
 
 	@Override
-	public void write(byte[] b, int off, int len) 
-		throws IOException 
-	{
+	public void write(byte[] b, int off, int len) throws IOException {
```

Comment:
The problem is this method, 

The rest of the PR is mainly just formatting and docs.

Reply:
Thanks, but could you also elaborate what the intended improvement is. From what I see you try to avoid a copy for tiny arrays, but it introduces brittle code that is prone to race conditions if the originating data stream flushes at certain points (danger that the array being currently written out to disk being overwritten by subsequent serialization).

Reply:
The code should not modify the handling of tiny arrays, if it does, I made a mistake.
The change is primarily only addressing the handling of byte[]'s of lengths that are longer than 8192. (or what we set out BYTE_BUFFER sizes to).

Where is the race condition?
The idea of the pool in this code is to force it into a single async stream that has one thread allocated to processing (as you designed it) the write task in seqence outside the 'hot' path.